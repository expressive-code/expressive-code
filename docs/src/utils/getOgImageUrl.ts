/**
 * Get the path to the OpenGraph image for a page
 */
export function getOgImageUrl(path: string): string | undefined {
	const imagePath = path.replace(/^\//, '').replace(/\/$/, '').replace(/^$/, 'index') + '.png'
	return '/open-graph/' + imagePath
}
