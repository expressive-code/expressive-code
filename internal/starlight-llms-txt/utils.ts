import type { CollectionEntry } from 'astro:content';
import { starlightLllmsTxtContext } from 'virtual:starlight-llms-txt/context';

const { defaultLocale, locales, title } = starlightLllmsTxtContext;
export const defaultLang = (defaultLocale === 'root' ? locales?.root?.lang : defaultLocale) || 'en';

/** Get the site title from the Starlight config. */
export function getSiteTitle(): string {
	return typeof title === 'string' ? title : (title[defaultLang] as string);
}

const localeKeys = Object.keys(locales || {}).filter(
	(key) => key !== 'root' && key !== defaultLang
);
const startsWithLocaleRE = new RegExp(`^(${localeKeys.join('|')})\/`);

/** Check if a content collection entry is part of the default locale or not. */
export function isDefaultLocale(doc: CollectionEntry<'docs'>): boolean {
	return !(localeKeys.includes(doc.id) || startsWithLocaleRE.test(doc.id));
}

/** Append a `/` to the passed string if it doesnâ€™t already end with one. */
export function ensureTrailingSlash(path: string) {
	return path.at(-1) === '/' ? path : path + '/';
}
