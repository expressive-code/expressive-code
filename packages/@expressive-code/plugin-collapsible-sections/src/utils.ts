import { ExpressiveCodeLine } from '@expressive-code/core'

export type Section = { from: number; to: number; lines: ExpressiveCodeLine[] }

/** Transforms a meta string of sections (e.g. '1-2, 4-8') into a list of section objects */
export function parseSections(value: string): Section[] {
	const sections: Section[] = []
	value
		.split(',')
		.map((section) => section.split('-').map((lineNum: string) => parseInt(lineNum)))
		.forEach((list) => {
			// skip any entries that don't have exactly 2 entries (e.g. '1-2-3' or '1-')
			if (list.length !== 2) return
			const [from, to] = list

			// skip any entries that failed to parse as numbers
			if (isNaN(from) || isNaN(to)) return

			// skip any entries that aren't increasing
			if (from > to) return

			// skip any entries that overlap existing sections, since our <details>-based approach can't overlap
			for (const { from: existingFrom, to: existingTo } of sections) {
				if (from >= existingFrom && from <= existingTo) return
				if (to >= existingFrom && to <= existingTo) return
			}

			sections.push({ from, to, lines: [] })
		})
	return sections
}
