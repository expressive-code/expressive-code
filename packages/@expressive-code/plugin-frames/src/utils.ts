export const frameTypes = ['code', 'terminal', 'none', 'auto'] as const

export type FrameType = (typeof frameTypes)[number]

/**
 * If the given input string represents a valid frame type,
 * converts it to a {@link FrameType} and returns it.
 *
 * Otherwise, returns `undefined`.
 */
export function frameTypeFromString(input: string) {
	// Support an empty string as alias for "none"
	if (input === '') input = 'none'

	// Fix common mistakes
	if (input === 'editor') input = 'code'
	if (input === 'shell') input = 'terminal'

	// Return either the converted type or undefined
	const frameType = input as FrameType
	return frameTypes.includes(frameType) ? frameType : undefined
}

export const LanguageGroups = {
	code: ['astro', 'cjs', 'htm', 'html', 'js', 'jsx', 'mjs', 'svelte', 'ts', 'tsx', 'typescript', 'vb', 'vue', 'vue-html'],
	terminal: ['bash', 'bat', 'batch', 'cmd', 'console', 'powershell', 'ps', 'ps1', 'psd1', 'psm1', 'sh', 'shell', 'shellscript', 'shellsession', 'zsh'],
	data: ['csv', 'env', 'ini', 'json', 'toml', 'xml', 'yaml', 'yml'],
	styles: ['css', 'less', 'sass', 'scss', 'styl', 'stylus', 'xsl'],
	textContent: ['markdown', 'md', 'mdx'],
}

export function isTerminalLanguage(language: string) {
	return LanguageGroups.terminal.includes(language)
}

const getFileNameCommentRegExpString = () =>
	[
		// Start of line
		`^`,
		// Optional whitespace
		`\\s*`,
		// Mandatory comment start: `//`, `#` (but not `#!`) or `<!--`
		`(?://|#(?!!)|<!--)`,
		// Optional whitespace
		`\\s*`,
		// Optional prefix before the file name:
		// - This is intended to match strings like `File name:` or `Example :`,
		//   but not Windows drive letters like `C:`,
		//   or URL protocols like `https:`
		// - We therefore expect the prefix to begin with any sequence of characters
		//   not starting with a letter + colon (to rule out Windows drive letters)
		// - The prefix must then be followed by:
		//   - a Japanese colon (`\\uff1a`), or
		//   - a regular colon (`:`) not followed by `//` (to rule out URL protocols)
		`(?:((?![a-z]:).*?)(?:\\uff1a|:(?!//)))?`,
		// Optional whitespace
		`\\s*`,
		// Capture the file name
		`(`,
		// Optional Windows drive letter
		`(?:[a-z]:)?`,
		// Optional sequence of characters allowed in file paths
		`[\\w./~%[\\]\\\\-]*`,
		// Optional dot and supported file extension
		`(?:\\.(?:${Object.values(LanguageGroups).flat().sort().join('|')}))?`,
		// End of file name capture
		`)`,
		// Optional whitespace
		`\\s*`,
		// Optional HTML comment end (`-->`)
		`(?:-->)?`,
		// Optional whitespace
		`\\s*`,
		// End of line
		`$`,
	].join('')

let fileNameCommentRegExp: RegExp | undefined

/**
 * Checks if the given source code line is a comment that contains a file name
 * for the code snippet.
 *
 * If the syntax highlighting language is contained in our known language groups,
 * only allows file names with extensions that belong to the same language group.
 */
export function getFileNameFromComment(line: string, lang: string) {
	if (fileNameCommentRegExp === undefined) {
		fileNameCommentRegExp = new RegExp(getFileNameCommentRegExpString(), 'i')
	}
	const matches = fileNameCommentRegExp.exec(line)
	const textBeforeFileName = matches?.[1] ?? ''
	const possibleFileName = matches?.[2]
	if (!possibleFileName) return

	// Check if the syntax highlighting language is contained in our known language groups,
	// and determine the extension of the extracted file name (if any)
	const languageGroup = Object.values(LanguageGroups).find((group) => group.includes(lang))
	const fileNameWithoutPath = possibleFileName.replace(/^.*[/\\]/, '')
	const fileExt = fileNameWithoutPath.match(/\.([^.]+)$/)?.[1]

	// Check if the file name has one or more typical file name patterns:
	// - It begins with any of these and is followed by at least one more character:
	//   - `/` (Unix path separator)
	//   - `\` (Windows path separator)
	//   - `./` or `.\`) (current directory)
	//   - `~` (home directory)
	//   - `[a-z]:` (Windows drive letter)
	// - It has a file name part starting with a dot (e.g. `some/path/.gitignore`)
	// - It looks like a separated path (see below for details)
	const hasTypicalFileNameBeginning = possibleFileName.match(/^(\/|\\|\.[/\\]|~|[a-z]:).+/i)
	const hasFileNameStartingWithDot = fileNameWithoutPath.startsWith('.')
	const looksLikeSeparatedPath =
		// Contains path separators
		possibleFileName.match(/[/\\]/) &&
		// Also contains other characters (except path separators, numbers and dots)
		possibleFileName.match(/[^/\\0-9.]/) &&
		// Does not contain spaces
		!possibleFileName.match(/\s/) &&
		// Is all lowercase
		possibleFileName === possibleFileName.toLowerCase()
	const hasTypicalFileNamePattern = hasTypicalFileNameBeginning || hasFileNameStartingWithDot || looksLikeSeparatedPath

	// Accept anything that looks like a file name if at least one of these conditions is true:
	// - the file name is the only text in the comment
	// - we're in the terminal language group (where extensions are often missing or unknown)
	if (hasTypicalFileNamePattern && (!textBeforeFileName.length || languageGroup === LanguageGroups.terminal)) {
		return possibleFileName
	}

	// Ignore the extracted file name if it doesn't have an extension,
	// or if its extension does not belong to the same language group
	// (e.g. JS code containing a CSS file name in a comment)
	if (!fileExt || (languageGroup && !languageGroup.includes(fileExt))) return

	return possibleFileName
}
