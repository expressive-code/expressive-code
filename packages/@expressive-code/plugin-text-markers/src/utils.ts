import { MarkerDefinition } from '.'

/**
 * Retrieves all group indices from the given RegExp match. Group indices are ranges
 * defined by start & end positions. The first group index refers to the full match,
 * and the following indices to RegExp capture groups (if any).
 *
 * If the RegExp flag `d` was enabled (and supported), it returns the native group indices.
 *
 * Otherwise, it uses fallback logic to manually search for the group contents inside the
 * full match. Note that this can be wrong if a group's contents can be found multiple times
 * inside the full match, but that's probably a rare case and still better than failing.
 */
export function getGroupIndicesFromRegExpMatch(match: RegExpMatchArray) {
	// Read the start and end ranges from the `indices` property,
	// which is made available through the RegExp flag `d`
	let groupIndices = match.indices as ([start: number, end: number] | null)[]
	if (groupIndices?.length) return groupIndices

	// We could not access native group indices, so we need to use fallback logic
	// to find the position of each capture group match inside the full match
	const fullMatchIndex = match.index as number
	groupIndices = match.map((groupValue) => {
		const groupIndex = groupValue ? match[0].indexOf(groupValue) : -1
		if (groupIndex === -1) return null
		const groupStart = fullMatchIndex + groupIndex
		const groupEnd = groupStart + groupValue.length
		return [groupStart, groupEnd]
	})

	return groupIndices
}

export function toDefinitionsArray(value: MarkerDefinition | MarkerDefinition[] | undefined) {
	if (value === undefined) return []
	return Array.isArray(value) ? value : [value]
}
