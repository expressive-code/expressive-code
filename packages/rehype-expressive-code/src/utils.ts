import type { Element } from 'expressive-code/hast'
import type { MdxJsxFlowElementHast, MdxJsxAttribute } from 'mdast-util-mdx-jsx'
import { getClassNames } from 'expressive-code/hast'

export type CodeBlockInfo = NonNullable<ReturnType<typeof getCodeBlockInfo>>

export function getCodeBlockInfo(pre: Element) {
	if (pre.tagName !== 'pre' || pre.children.length !== 1) return
	const code = pre.children[0]
	if (code.type !== 'element' || code.tagName !== 'code') return
	const text = code.children[0]
	if (text.type !== 'text') return
	const langClass = getClassNames(code).find((c) => c.startsWith('language-')) ?? ''
	const lang = langClass.replace('language-', '')
	const meta = (code.data?.meta ?? code.properties?.metastring ?? '') as string
	return {
		pre,
		code,
		lang,
		text: text.value,
		meta,
	}
}

export function createInlineAssetElement({
	tagName,
	properties = {},
	innerHTML,
	useMdxJsx,
}: {
	tagName: string
	properties?: Record<string, string | null | undefined> | undefined
	innerHTML: string
	useMdxJsx: boolean
}): MdxJsxFlowElementHast | Element {
	if (useMdxJsx) return createMdxJsxElementWithInnerHTML(tagName, properties, innerHTML)
	return {
		type: 'element',
		tagName,
		properties,
		children: [{ type: 'text', value: innerHTML }],
	}
}

function createMdxJsxElementWithInnerHTML(tagName: string, properties: Record<string, MdxJsxAttribute['value']>, innerHTML: string): MdxJsxFlowElementHast {
	const attrs: MdxJsxAttribute[] = Object.entries(properties).map(([name, value]: [string, MdxJsxAttribute['value']]) => ({
		type: 'mdxJsxAttribute',
		name,
		value,
	}))
	return {
		type: 'mdxJsxFlowElement',
		name: tagName,
		attributes: [
			...attrs,
			{
				type: 'mdxJsxAttribute',
				name: 'dangerouslySetInnerHTML',
				value: {
					type: 'mdxJsxAttributeValueExpression',
					value: `{__html:${JSON.stringify(innerHTML)}}`,
					data: {
						estree: {
							type: 'Program',
							body: [
								{
									type: 'ExpressionStatement',
									expression: {
										type: 'ObjectExpression',
										properties: [
											{
												type: 'Property',
												method: false,
												shorthand: false,
												computed: false,
												key: {
													type: 'Identifier',
													name: '__html',
												},
												value: {
													type: 'Literal',
													value: innerHTML,
													raw: JSON.stringify(innerHTML),
												},
												kind: 'init',
											},
										],
									},
								},
							],
							sourceType: 'module',
							comments: [],
						},
					},
				},
			},
		],
		data: {
			_mdxExplicitJsx: true,
		},
		children: [],
	}
}
