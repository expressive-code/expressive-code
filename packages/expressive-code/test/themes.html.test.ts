import { beforeAll, describe, expect, test } from 'vitest'
import { addPrettierHtmlSnapshotSerializer, getAnnotationResult } from './utils'

const codeSnippet = `
---
import MyReactComponent from '../components/MyReactComponent.jsx';
import MyAstroComponent from '../components/MyAstroComponent.astro';
---

<MyReactComponent>
	<MyAstroComponent slot="name" />
</MyReactComponent>
`

beforeAll(() => {
	addPrettierHtmlSnapshotSerializer()
})

describe('Color contrast', () => {
	test('Lightens tokens inside marked lines', () => {
		const annotationResult = getAnnotationResult(codeSnippet, {
			annotations: {
				lineMarkings: [{ lines: [2, 7] }],
			},
		})

		expect(annotationResult.annotatedCodeHtml).toMatchSnapshot()
	})
})

// TODO: Add color contrast tests, including cases where color must be inverted, and where contrast cannot be improved
