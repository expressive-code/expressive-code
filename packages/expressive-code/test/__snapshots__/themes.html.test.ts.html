// Vitest Snapshot v1

exports[`Color contrast > Lightens tokens inside marked lines 1`] = `
<pre class="shiki" style="background-color: #2e3440ff; color: #d8dee9ff"
	><div class="language-id">astro</div
	><div class="code-container"
		><code
			><div class="line"><span style="color: #616e88">---</span></div
			><div class="line mark"
				><span style="color: #a6bdd3">import</span
				><span style="color: #d8dee9"> </span
				><span style="color: #d8dee9">MyReactComponent</span
				><span style="color: #d8dee9"> </span
				><span style="color: #a6bdd3">from</span
				><span style="color: #d8dee9"> </span
				><span style="color: #eceff4">'</span
				><span style="color: #a9c293">../components/MyReactComponent.jsx</span
				><span style="color: #eceff4">'</span
				><span style="color: #a6bdd3">;</span></div
			><div class="line"
				><span style="color: #81a1c1">import</span
				><span style="color: #d8dee9ff"> </span
				><span style="color: #d8dee9">MyAstroComponent</span
				><span style="color: #d8dee9ff"> </span
				><span style="color: #81a1c1">from</span
				><span style="color: #d8dee9ff"> </span
				><span style="color: #eceff4">'</span
				><span style="color: #a3be8c">../components/MyAstroComponent.astro</span
				><span style="color: #eceff4">'</span
				><span style="color: #81a1c1">;</span></div
			><div class="line"><span style="color: #616e88">---</span></div
			><div class="line"></div
			><div class="line"
				><span style="color: #81a1c1">&lt;</span
				><span style="color: #8fbcbb">MyReactComponent</span
				><span style="color: #81a1c1">&gt;</span></div
			><div class="line mark"
				><span style="color: #d8dee9"> </span
				><span style="color: #a6bdd3">&lt;</span
				><span style="color: #99c2c1">MyAstroComponent</span
				><span style="color: #d8dee9"> </span
				><span style="color: #99c2c1">slot</span
				><span style="color: #eceff4">=</span
				><span style="color: #eceff4">"</span
				><span style="color: #a9c293">name</span
				><span style="color: #eceff4">"</span
				><span style="color: #d8dee9"> </span
				><span style="color: #a6bdd3">/&gt;</span></div
			><div class="line"
				><span style="color: #81a1c1">&lt;/</span
				><span style="color: #8fbcbb">MyReactComponent</span
				><span style="color: #81a1c1">&gt;</span></div
			></code
		></div
	></pre
>

`;
